### PR TITLE
Replay tool asynchronous stream dispatch

### PR DESCRIPTION
**Pull request by benouaer**

### CODE REVIEW

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java [hunk 1]
```diff
@@ -1,60 +1,75 @@
...
-
+    private boolean allTasksCreated = false;
 
     @PostConstruct
     void go() throws InterruptedException {
-        final long startTimeMillis = currentTimeMillis();
-        LOGGER.info("-------------- Replay Event Streams -------------!");
+        LOGGER.info("-------------- Invoke Event Streams Replay-------------!");
+        jdbcEventRepository.getStreamOfAllEventStreams()
+                .forEach(stream -> {
```
Comment (skiddykong) [inline framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: eventStream?

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java [hunk 2]
```diff
@@ -1,60 +1,75 @@
...
+        boolean dispatchComplete = removeOutstandingTask(dispatchTaskFuture) && allTasksCreated;
+        if (dispatchComplete) shutdown();
+    }
 
-            while (!finished()) {
-                Thread.sleep(3000);
-            }
-        }
+    public void taskDone(Future<?> dispatchTaskFuture, ManagedExecutorService managedExecutorService, Object dispatchTask, Throwable throwable) {
+        boolean dispatchComplete = removeOutstandingTask(dispatchTaskFuture) && allTasksCreated;
```
Comment (skiddykong) [inline framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: final

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java [hunk 3]
```diff
@@ -1,60 +1,75 @@
...
+                });
+        allTasksCreated = true;
+        if (outstandingTasks.isEmpty()) shutdown();
+        LOGGER.info("-------------- Invocation of Event Streams Replay Completed --------------");
+    }
 
-        try (final Stream<Stream<JsonEnvelope>> streamOfEventStreams = jdbcEventRepository.getStreamOfAllEventStreams()) {
-            streamOfEventStreams.forEach(s -> dispatchResults.add(asyncStreamDispatcher.dispatch(s)));
-            started = true;
+    public void taskAborted(Future<?> dispatchTaskFuture, ManagedExecutorService managedExecutorService, Object dispatchTask, Throwable throwable) {
```
Comment (skiddykong) [inline framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: parameters should be final

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java [hunk 4]
```diff
@@ -1,60 +1,75 @@
...
+        boolean dispatchComplete = removeOutstandingTask(dispatchTaskFuture) && allTasksCreated;
+        if (dispatchComplete) shutdown();
+    }
 
-            while (!finished()) {
-                Thread.sleep(3000);
-            }
-        }
+    public void taskDone(Future<?> dispatchTaskFuture, ManagedExecutorService managedExecutorService, Object dispatchTask, Throwable throwable) {
```
Comment (skiddykong) [inline framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: final

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java [hunk 5]
```diff
@@ -1,60 +1,75 @@
...
+        boolean dispatchComplete = removeOutstandingTask(dispatchTaskFuture) && allTasksCreated;
+        if (dispatchComplete) shutdown();
+    }
 
-        LOGGER.info("-------------- Replay of Event Streams Completed in {} milliseconds--------------", currentTimeMillis() - startTimeMillis);
+    public void taskStarting(Future<?> dispatchTaskFuture, ManagedExecutorService managedExecutorService, Object dispatchTask) {
```
Comment (skiddykong) [inline framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: finals for parameters

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java [hunk 6]
```diff
@@ -1,60 +1,75 @@
...
+        LOGGER.debug("Starting Dispatch task");
     }
 
+    public void taskSubmitted(Future<?> dispatchTaskFuture, ManagedExecutorService managedExecutorService, Object dispatchTask) {
```
Comment (skiddykong) [inline framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: finals for parameters

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java [hunk 7]
```diff
@@ -1,60 +1,75 @@
...
+        LOGGER.debug("Submitted Dispatch task");
+    }
+
+    private boolean removeOutstandingTask(Future<?> dispatchTaskFuture) {
```
Comment (skiddykong) [inline framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: finals for parameters

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StreamDispatchTask.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import javax.enterprise.concurrent.ManagedTaskListener;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.stream.Stream;
+
+
+public class StreamDispatchTask implements Callable<UUID>, ManagedTask {
+    private static final Logger LOGGER = LoggerFactory.getLogger(StreamDispatchTask.class);
+    private Stream<JsonEnvelope> stream;
```
Comment (skiddykong) [inline framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StreamDispatchTask.java]: final

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StreamDispatchTask.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    private AsyncStreamDispatcher dispatcher;
```
Comment (skiddykong) [inline framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StreamDispatchTask.java]: final

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StreamDispatchTask.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    private ManagedTaskListener dispatchListener;
```
Comment (skiddykong) [inline framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StreamDispatchTask.java]: final

File: framework-tools-replay/src/test/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcherTest.java [hunk 11]
```diff
@@ -50,6 +47,9 @@ public void shouldDispatchEnvelopes() throws Exception {
         final JsonEnvelope envelope1 = envelope().with(metadataWithDefaults().withStreamId(streamId).withVersion(1L)).build();
         final JsonEnvelope envelope2 = envelope().with(metadataWithDefaults().withStreamId(streamId).withVersion(2L)).build();
 
+        doNothing().when(envelopeDispatcher).dispatch(envelope1);
+        doNothing().when(envelopeDispatcher).dispatch(envelope2);
+
         asyncStreamDispatcher.dispatch(Stream.of(envelope1, envelope2));
 
         ArgumentCaptor<JsonEnvelope> dispatchCaptor = ArgumentCaptor.forClass(JsonEnvelope.class);
```
Comment (skiddykong) [inline framework-tools-replay/src/test/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcherTest.java]: final

File: framework-tools-replay/src/test/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcherTest.java [hunk 12]
```diff
@@ -74,16 +74,14 @@ public void shouldUpdateStreamBufferStatus() throws Exception {
 
     @Test(expected = IllegalArgumentException.class)
     public void shouldThrowExceptionIfNoStreamIdInTheEnvelope() {
-
-        asyncStreamDispatcher.dispatch(Stream.of(envelope().with(metadataWithRandomUUID("dummyName").withVersion(1L)).build()));
-
+        Stream<JsonEnvelope> stream = Stream.of(envelope().with(metadataWithRandomUUID("dummyName").withVersion(1L)).build());
```
Comment (skiddykong) [inline framework-tools-replay/src/test/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcherTest.java]: final

File: framework-tools-replay/src/test/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcherTest.java [hunk 13]
```diff
@@ -74,16 +74,14 @@ public void shouldUpdateStreamBufferStatus() throws Exception {
...
+        asyncStreamDispatcher.dispatch(stream);
     }
 
     @Test(expected = IllegalArgumentException.class)
     public void shouldThrowExceptionIfNoVersionInTheEnvelope() {
-
-        asyncStreamDispatcher.dispatch(Stream.of(envelope().with(metadataWithDefaults().withStreamId(randomUUID())).build()));
-
+        Stream<JsonEnvelope> stream = Stream.of(envelope().with(metadataWithDefaults().withStreamId(randomUUID())).build());
```
Comment (skiddykong) [inline framework-tools-replay/src/test/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcherTest.java]: final

File: pom.xml [hunk 14]
```diff
@@ -31,7 +29,6 @@
     <modules>
         <module>framework-tools-command</module>
         <module>framework-tools-replay</module>
-        <module>framework-tools-test</module>
```
Comment (skiddykong) [inline pom.xml]: Why has the IT been deleted?!!!

File: framework-tools-replay/pom.xml [hunk 15]
```diff
@@ -112,6 +112,12 @@
             <groupId>org.slf4j</groupId>
             <artifactId>slf4j-log4j12</artifactId>
         </dependency>
+        <dependency>
+            <groupId>javax</groupId>
+            <artifactId>javaee-api</artifactId>
+            <version>7.0</version>
```
[the remark this answers is not in the document]
Comment (benouaer) [reply framework-tools-replay/pom.xml]: moved

File: framework-tools-replay/pom.xml [hunk 16]
```diff
@@ -112,6 +112,12 @@
...
+            <!--<scope>provided</scope>-->
```
[the remark this answers is not in the document]
Comment (benouaer) [reply framework-tools-replay/pom.xml]: removed

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcher.java [hunk 17]
```diff
@@ -1,46 +1,38 @@
...
-
+    private TransactionalEnvelopeDispatcher envelopeDispatcher;
     @Inject
     private StreamStatusJdbcRepository streamStatusRepository;
 
-    @Asynchronous
-    public Future<Void> dispatch(final Stream<JsonEnvelope> stream) {
+    public UUID dispatch(final Stream<JsonEnvelope> stream) {
         final int[] noOfProcessedElements = {0};
         final JsonEnvelope[] envelope = {null};
```
[the remark this answers is not in the document]
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcher.java]: renamed

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcher.java [hunk 18]
```diff
@@ -1,46 +1,38 @@
...
+        UUID streamId;
```
[the remark this answers is not in the document]
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcher.java]: done

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcher.java [hunk 19]
```diff
@@ -1,46 +1,38 @@
...
 
         try (final Stream<JsonEnvelope> stream1 = stream) {
             stream1.forEach(e -> {
```
[the remark this answers is not in the document]
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcher.java]: renamed

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java (hunk 1 above)
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: renamed

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java (hunk 3 above)
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: done

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java (hunk 4 above)
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: done

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java (hunk 2 above)
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: done

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java (hunk 5 above)
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: done

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java (hunk 6 above)
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: done

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java (hunk 7 above)
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StartReplay.java]: done

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StreamDispatchTask.java (hunk 8 above)
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StreamDispatchTask.java]: done

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StreamDispatchTask.java (hunk 9 above)
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StreamDispatchTask.java]: done

File: framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StreamDispatchTask.java (hunk 10 above)
Comment (benouaer) [reply framework-tools-replay/src/main/java/uk/gov/justice/framework/tools/replay/StreamDispatchTask.java]: done

File: framework-tools-replay/src/test/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcherTest.java (hunk 11 above)
Comment (benouaer) [reply framework-tools-replay/src/test/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcherTest.java]: done

File: framework-tools-replay/src/test/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcherTest.java (hunk 13 above)
Comment (benouaer) [reply framework-tools-replay/src/test/java/uk/gov/justice/framework/tools/replay/AsyncStreamDispatcherTest.java]: done

File: pom.xml (hunk 14 above)
Comment (benouaer) [reply pom.xml]: reinstated

File: framework-tools-replay/pom.xml [hunk 20]
```diff
@@ -112,6 +112,11 @@
             <groupId>org.slf4j</groupId>
             <artifactId>slf4j-log4j12</artifactId>
         </dependency>
+        <dependency>
+            <groupId>javax</groupId>
+            <artifactId>javaee-api</artifactId>
+            <version>${java.ee.version}</version>
```
Comment (skiddykong) [inline framework-tools-replay/pom.xml]: not sure you need this, dependency is managed

File: pom.xml [hunk 21]
```diff
@@ -20,12 +20,12 @@
         <test.utils.core.version>1.15.0</test.utils.core.version>
         <cpp.repo.name>framework-tools</cpp.repo.name>
         <uk.gov.justice.utils.version>1.5.0</uk.gov.justice.utils.version>
-
         <!--
                 note h2 version is because swarm has its own version of h2 which we must use
                 for the integration tests to work as h2 version inter-compatibility is volatile
         -->
         <com.h2database.version>1.4.187</com.h2database.version>
+        <java.ee.version>7.0</java.ee.version>
```
Comment (skiddykong) [inline pom.xml]: this is managed in the MS pom I think

**Pull request merged**